# Real estate prices analysis and prediction
#### Prepared by: 
- Ahmed
- Hatem
- Nesma

#### Introduction
(Nesma)

In Egypt, real-estate trade is one of the major investment methods that people usually choose to protect the value of their money savings. Egyptians consider it a robust way to protect their money value against currency inflation. In many cities, people even take pride in being owners of houses.


In this project, we decided to dig deep in the real estate market to analyze the data available and answer some questions we had regarding the real-estate market in Egypt, below are our questions:



1. What are the most important factors that affect the house price?

2. Seasonality behavior. Is there certain months that have higher sales trend over the other months?

3. Are there certain areas that people prefer to buy houses in?

4. Do Egyptians prefer luxious houses or economic houses?

#### The Dataset
(Nesma)

We decided to retrieve the required data from the famous Egyptian real estate ads website "aqarmap" that people use to buy and sell real-estate. Because it is widely known among Egyptians and we thought that this means it will include enough data that can help us answer our questions (There are 2 Million consumers that use Aqarmap every month according to the linked source:
https://i.aqarmap.com/trends/aqarmap-trends-2019-en.pdf).

To retrieve the data, I used the readily available Web Scraping tool downloaded from "https://www.parsehub.com/".
Our team (Ahmed, Hatem and me) agreed to limit our analysis to apartments in Greater Cairo only, because it has dense population as compared to other governorates. I sorted the data by "Newest First" to work on fresh recent records.

After refining the search criteria in the website, it showed that there are 75,702 ads that meets this criteria.

!["Drawing"](P4.png)

However, the scraping tool returned around 300 records only. By checking the results returned from the website and using the "Next" button, I found that the last page was #20. Knowing that each page contains 15 ad, then the data is around 300. which nearly matches the data retrieved from scraping.

!["Drawing"](P3.png)

I tried another search that included all governorates and it also returned 20 pages only, which means that this is the maximum number of pages the site displays for any search.

I wanted to add more data, so I made another scraping after adding "automatic sort" to the search.

let's now check the data!

#### Checking the data
(Nesma)

* Please note that the scrapped data was returned in 3 separate files.
* While using the scraping tool, I selected the summary of each ad from the outer page, and then opended each ad to get its URL and details. In this analysis, We will focus on the details, while the summary won't be important.

!["Drawing"](p5.png)

In [1]:
# Let's start by importing all libraries used in our analysis

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
# Now let's create a dataset from the first file and take a look at the first 2 rows

df_a= pd.read_csv('1_run_results.csv')

df_a.head(2)

Summary_name  \
0  شقة للبيع مكان مميز جدااا\n6th of October, Gre...   
1  شقه للبيع في حي 12مج4 أكتوبر\n6th of October, ...   

                                         Summary_url Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133778-for-...                4   
1  https://aqarmap.com.eg/en/listing/3133773-for-...                3   

  Summary_No_Rooms Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
0                4                    2          245 M²     EG-3133778   
1                2                    1           84 M²     EG-3133773   

  Summary_Publishing_Date Summary_Price_EGP Summary_Seller_Type  \
0              06/01/2022     2,000,000 EGP      Property Owner   
1              06/01/2022       400,000 EGP      Property Owner   

  Summary_Price_P_M  
0      8,163 EGP/M²  
1      4,762 EGP/M²

In [3]:
# Next, we will create a dataset from the second file

df_b= pd.read_csv('2_run_results.csv')

df_b.head(2)

Summary_name  \
0  Corner Penthouse Apartment in Cairo University...   
1  Apartments For sale in Dar Misr El Koronfel\nد...   

                                         Summary_url Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133147-for-...                4   
1  https://aqarmap.com.eg/en/listing/3133145-for-...                4   

   Summary_No_Rooms Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
0               4.0                    4            2022  SEMI_FINISHED   
1               2.0                    2            2022            LUX   

  Summary_Publishing_Date Summary_Price_EGP  \
0                  Corner            350 M²   
1                   Other            130 M²   

                           Summary_Seller_Type Summary_Price_P_M  
0                                   EG-3133147          5/1/2022  
1  Apartments For sale in Dar Misr El Koronfel          5/1/2022

In [4]:
# Finally, we will create a dataset from the third file

df_c= pd.read_csv('3_run_results.csv')

df_c.head(2)

Summary_name  \
0  Apartments For sale in Second Neighborhood\nال...   
1  Apartments For sale in Investors Area\nمنطقة ا...   

                                         Summary_url Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3126749-for-...                8   
1  https://aqarmap.com.eg/en/listing/3102080-for-...                3   

  Summary_No_Rooms Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
0                2                    1            2024  SEMI_FINISHED   
1                3                    3   SEMI_FINISHED         Garden   

  Summary_Publishing_Date Summary_Price_EGP  \
0             Main Street            100 M²   
1                  173 M²        EG-3102080   

                          Summary_Seller_Type Summary_Price_P_M  
0  Apartments For sale in Second Neighborhood        02/01/2022  
1                                  16/12/2021     2,000,000 EGP

Since the same tool was used to retrieve all files, then we are confident that the columns ordering is the same
in all of them. Now let's merge all datasets.

In [5]:
df_new= pd.concat([df_a,df_b, df_c])
len(df_new)

436

In [6]:
len(df_a)

188

In [7]:
len(df_b)

60

In [8]:
len(df_c)

188

Now that our dataset is available, let's start our cleaning!

#### Preprocessing the data
(Nesma)

In [9]:
df_new.head()

Summary_name  \
0  شقة للبيع مكان مميز جدااا\n6th of October, Gre...   
1  شقه للبيع في حي 12مج4 أكتوبر\n6th of October, ...   
2  شقه الترا لوكس\nHamad Agami St., Kafr Tuhurmos...   
3  شقة للبيع\nNasr City, Greater Cairo\n101 M² 4 ...   
4  شقة للبيع في مدينة العبور\nCompounds in El Obo...   

                                         Summary_url Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133778-for-...                4   
1  https://aqarmap.com.eg/en/listing/3133773-for-...                3   
2  https://aqarmap.com.eg/en/listing/3133772-for-...                5   
3  https://aqarmap.com.eg/en/listing/3133770-for-...                3   
4  https://aqarmap.com.eg/en/listing/3133768-for-...                2   

  Summary_No_Rooms Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
0                4                    2          245 M²     EG-3133778   
1                2                    1           84 M²     EG-3133773   
2                3                    1          110 M²     EG-3133772   
3                4                    1          101 M²     EG-3133770   
4                1                68 M²      EG-3133768     06/01/2022   

  Summary_Publishing_Date Summary_Price_EGP Summary_Seller_Type  \
0              06/01/2022     2,000,000 EGP      Property Owner   
1              06/01/2022       400,000 EGP      Property Owner   
2              06/01/2022       260,000 EGP      شقه الترا لوكس   
3              06/01/2022       650,000 EGP      Property Owner   
4             315,000 EGP    Property Owner        4,632 EGP/M²   

  Summary_Price_P_M  
0      8,163 EGP/M²  
1      4,762 EGP/M²  
2      2,364 EGP/M²  
3      6,436 EGP/M²  
4               NaN

Let's start by removing the duplicate rows. We can use the "Summary_url" column to remove the duplication because there will never be two ads with the same exact URL

In [10]:
df=df_new.drop_duplicates(subset='Summary_url')
df.head()

Summary_name  \
0  شقة للبيع مكان مميز جدااا\n6th of October, Gre...   
1  شقه للبيع في حي 12مج4 أكتوبر\n6th of October, ...   
2  شقه الترا لوكس\nHamad Agami St., Kafr Tuhurmos...   
3  شقة للبيع\nNasr City, Greater Cairo\n101 M² 4 ...   
4  شقة للبيع في مدينة العبور\nCompounds in El Obo...   

                                         Summary_url Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133778-for-...                4   
1  https://aqarmap.com.eg/en/listing/3133773-for-...                3   
2  https://aqarmap.com.eg/en/listing/3133772-for-...                5   
3  https://aqarmap.com.eg/en/listing/3133770-for-...                3   
4  https://aqarmap.com.eg/en/listing/3133768-for-...                2   

  Summary_No_Rooms Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
0                4                    2          245 M²     EG-3133778   
1                2                    1           84 M²     EG-3133773   
2                3                    1          110 M²     EG-3133772   
3                4                    1          101 M²     EG-3133770   
4                1                68 M²      EG-3133768     06/01/2022   

  Summary_Publishing_Date Summary_Price_EGP Summary_Seller_Type  \
0              06/01/2022     2,000,000 EGP      Property Owner   
1              06/01/2022       400,000 EGP      Property Owner   
2              06/01/2022       260,000 EGP      شقه الترا لوكس   
3              06/01/2022       650,000 EGP      Property Owner   
4             315,000 EGP    Property Owner        4,632 EGP/M²   

  Summary_Price_P_M  
0      8,163 EGP/M²  
1      4,762 EGP/M²  
2      2,364 EGP/M²  
3      6,436 EGP/M²  
4               NaN

In [11]:
len(df)

428

Now, let's check the data information to see if any column has a value that doesn't match the remaining values

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 428 entries, 0 to 187
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Summary_name             428 non-null    object
 1   Summary_url              428 non-null    object
 2   Summary_Floor_No         426 non-null    object
 3   Summary_No_Rooms         426 non-null    object
 4   Summary_No_Bathrooms     426 non-null    object
 5   Summary_Area_m2          426 non-null    object
 6   Summary_Add_ID           426 non-null    object
 7   Summary_Publishing_Date  426 non-null    object
 8   Summary_Price_EGP        426 non-null    object
 9   Summary_Seller_Type      426 non-null    object
 10  Summary_Price_P_M        412 non-null    object
dtypes: object(11)
memory usage: 40.1+ KB


#### Two things to notice here:
1. We need to get rid of any row that has missing or NA values across all the columns we are interested in.
2. Although the majority of columns should include integers only, but it seems that they include string values which caused the data type to be  "object". So we need to check each column individually.

In [13]:
# Let's drop the columns with NA values
df1=df.dropna(axis=0, subset=df.columns[2:])
df1.head(2)

Summary_name  \
0  شقة للبيع مكان مميز جدااا\n6th of October, Gre...   
1  شقه للبيع في حي 12مج4 أكتوبر\n6th of October, ...   

                                         Summary_url Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133778-for-...                4   
1  https://aqarmap.com.eg/en/listing/3133773-for-...                3   

  Summary_No_Rooms Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
0                4                    2          245 M²     EG-3133778   
1                2                    1           84 M²     EG-3133773   

  Summary_Publishing_Date Summary_Price_EGP Summary_Seller_Type  \
0              06/01/2022     2,000,000 EGP      Property Owner   
1              06/01/2022       400,000 EGP      Property Owner   

  Summary_Price_P_M  
0      8,163 EGP/M²  
1      4,762 EGP/M²

In [14]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 412 entries, 0 to 186
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Summary_name             412 non-null    object
 1   Summary_url              412 non-null    object
 2   Summary_Floor_No         412 non-null    object
 3   Summary_No_Rooms         412 non-null    object
 4   Summary_No_Bathrooms     412 non-null    object
 5   Summary_Area_m2          412 non-null    object
 6   Summary_Add_ID           412 non-null    object
 7   Summary_Publishing_Date  412 non-null    object
 8   Summary_Price_EGP        412 non-null    object
 9   Summary_Seller_Type      412 non-null    object
 10  Summary_Price_P_M        412 non-null    object
dtypes: object(11)
memory usage: 38.6+ KB


Now we need to work on each column. Let's start by checking the "Summary_Floor_No" column

In [15]:
df1['Summary_Floor_No'].unique()

array(['4', '3', '5', '2', '8', '1', '9', '7', '6', 'SEMI_FINISHED',
       'Ground', '10', '17', '11', '12', 'extra_super_lux',
       'semi_finished'], dtype=object)

We want to specify the rows with values of floor that are not-numeric. We can do this by sorting the values.

In [16]:
df2=df1.sort_values(by="Summary_Floor_No")
df2.tail(5)

Summary_name  \
138  Duplex 161 M² Super lux in Monte Napoleone Com...   
37   Own your apartment in New Capital -discount of...   
22   شقة 185 متر في كمبوند حياتي ريزيدنس نص تشطيب ب...   
18   Direct from Owner\nشقة تشطيب سوبر لوكس فى النر...   
165  شقة ٢٥٥م بالحي الثالث شرق ع الاوسط مباشرة\nNei...   

                                           Summary_url Summary_Floor_No  \
138  https://aqarmap.com.eg/en/listing/3096106-for-...           Ground   
37   https://aqarmap.com.eg/en/listing/3133072-for-...           Ground   
22   https://aqarmap.com.eg/en/listing/3133671-for-...    SEMI_FINISHED   
18   https://aqarmap.com.eg/en/listing/3128718-for-...  extra_super_lux   
165  https://aqarmap.com.eg/en/listing/3090051-for-...    semi_finished   

    Summary_No_Rooms Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
138                2                    1            2025      SUPER_LUX   
37               2.0                    2            2021  SEMI_FINISHED   
22       Side Street               185 M²      EG-3133671     05/01/2022   
18       Side Street               200 M²      EG-3128718     03/01/2022   
165      Main Street               255 M²      EG-3090051     08/12/2021   

    Summary_Publishing_Date Summary_Price_EGP  \
138                  Garden            161 M²   
37                   Garden            127 M²   
22            2,000,000 EGP             Agent   
18            2,800,000 EGP    Property Owner   
165           1,350,000 EGP             Agent   

                                   Summary_Seller_Type     Summary_Price_P_M  
138  Duplex 161 M² Super lux in Monte Napoleone Com...            12/12/2021  
37   Own your apartment in New Capital -discount of...              5/1/2022  
22   شقة 185 متر في كمبوند حياتي ريزيدنس نص تشطيب ب...         10,811 EGP/M²  
18                     شقة تشطيب سوبر لوكس فى النرجس ١  Commercial Financing  
165          شقة ٢٥٥م بالحي الثالث شرق ع الاوسط مباشرة          5,294 EGP/M²

For the last 3 rows, the data doesn't include information about the rooms or bathrooms, so we can drop them

In [17]:
df3=df2[:-3]
df3.tail(2)

Summary_name  \
138  Duplex 161 M² Super lux in Monte Napoleone Com...   
37   Own your apartment in New Capital -discount of...   

                                           Summary_url Summary_Floor_No  \
138  https://aqarmap.com.eg/en/listing/3096106-for-...           Ground   
37   https://aqarmap.com.eg/en/listing/3133072-for-...           Ground   

    Summary_No_Rooms Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
138                2                    1            2025      SUPER_LUX   
37               2.0                    2            2021  SEMI_FINISHED   

    Summary_Publishing_Date Summary_Price_EGP  \
138                  Garden            161 M²   
37                   Garden            127 M²   

                                   Summary_Seller_Type Summary_Price_P_M  
138  Duplex 161 M² Super lux in Monte Napoleone Com...        12/12/2021  
37   Own your apartment in New Capital -discount of...          5/1/2022

We notice that the ground floor is written as a text, We need to replace it with 0 to match the rest of column values

In [18]:
df3['Summary_Floor_No'].replace(to_replace= "Ground", value= 0, inplace=True)
df3.tail(2)

D:\Anaconda\lib\site-packages\pandas\core\series.py:4509: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().replace(


Summary_name  \
138  Duplex 161 M² Super lux in Monte Napoleone Com...   
37   Own your apartment in New Capital -discount of...   

                                           Summary_url Summary_Floor_No  \
138  https://aqarmap.com.eg/en/listing/3096106-for-...                0   
37   https://aqarmap.com.eg/en/listing/3133072-for-...                0   

    Summary_No_Rooms Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
138                2                    1            2025      SUPER_LUX   
37               2.0                    2            2021  SEMI_FINISHED   

    Summary_Publishing_Date Summary_Price_EGP  \
138                  Garden            161 M²   
37                   Garden            127 M²   

                                   Summary_Seller_Type Summary_Price_P_M  
138  Duplex 161 M² Super lux in Monte Napoleone Com...        12/12/2021  
37   Own your apartment in New Capital -discount of...          5/1/2022

We also need to change the format of floor numbers to be integer not string

In [19]:
df3['Summary_Floor_No']=df3['Summary_Floor_No'].astype('int')
df3['Summary_Floor_No'].unique()

<ipython-input-19-5245d50ecaed>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df3['Summary_Floor_No']=df3['Summary_Floor_No'].astype('int')


array([ 1, 10, 11, 12, 17,  2,  3,  4,  5,  6,  7,  8,  9,  0])

Now, let's move to the "Summary_No_Rooms" column

In [20]:
df3['Summary_No_Rooms'].unique()

array(['3', 4.0, '2', '4', 3.0, '5', 2.0, '6', '1', 5.0], dtype=object)

The values make sense, but let's also cast the type to integer

In [21]:
df3['Summary_No_Rooms']=df3['Summary_No_Rooms'].astype('int')
df3['Summary_No_Rooms'].unique()

<ipython-input-21-aea05c527830>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df3['Summary_No_Rooms']=df3['Summary_No_Rooms'].astype('int')


array([3, 4, 2, 5, 6, 1])

Next, let's move on to the "Summary_No_Bathrooms" column

In [22]:
df3['Summary_No_Bathrooms'].unique()

array(['1', '4', '2', '3', '6', '5'], dtype=object)

let's cast it to integer

In [23]:
df3['Summary_No_Bathrooms']=df3['Summary_No_Bathrooms'].astype('int')
df3['Summary_No_Bathrooms'].unique()

<ipython-input-23-04656743e834>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df3['Summary_No_Bathrooms']=df3['Summary_No_Bathrooms'].astype('int')


array([1, 4, 2, 3, 6, 5])

Next, we have the "Summary_Area_m2" column. let's check the values

In [24]:
df3['Summary_Area_m2'].unique()

array(['2023', '2021', '2020', '2024', '2018', '2022', '2019', '500 M²',
       '2014', 'SEMI_FINISHED', '145 M²', '120 M²', '150 M²', '170 M²',
       '2015', '90 M²', '140 M²', '250 M²', '115 M²', '200 M²', '180 M²',
       '2009', '2017', '2025', '330 M²', '1980', '2010', '195 M²', '1995',
       '1990', '110 M²', '156 M²', '240 M²', '95 M²', 'EXTRA_SUPER_LUX',
       '130 M²', '63 M²', 'SUPER_LUX', '530 M²', '175 M²', '2013',
       '125 M²', '2008', '80 M²', '183 M²', '84 M²', '190 M²', '2016',
       '56 M²', '2011', '101 M²', '1999', '1988', '2007', '245 M²',
       '136 M²', '57 M²', '202 M²', '75 M²', '164 M²', '100 M²', '220 M²',
       '96 M²', '70 M²', '139 M²', '1970', '2000', '265 M²', '65 M²',
       '105 M²', '2006', '185 M²', '160 M²', '300 M²', '165 M²', '1985'],
      dtype=object)

In [25]:
df3['Summary_Area_m2'].value_counts()[:20]

2022             67
2021             58
2023             34
2020             30
2024             28
2019             16
2018             15
2015             13
140 M²            7
110 M²            6
130 M²            6
150 M²            6
SUPER_LUX         6
120 M²            6
2025              5
2017              5
2014              4
2000              4
SEMI_FINISHED     4
90 M²             4
Name: Summary_Area_m2, dtype: int64

By checking the data, it seems that the records in which the finishing type is mentioned are a small number, so we can ignore these values. The majority of values here belong to a number that seems to be a year. However, it's not clear whether this is the construction year or the delivery year. So I will also ignore it, and will replace both values by the area, which already exists in these rows but shifted to the right.

In [26]:
# Let's first sort the data by this column
df4= df3.sort_values(by="Summary_Area_m2").reset_index()
df4.tail(5)

index                                       Summary_name  \
404    164  Apartment For sale,170M @90th Avenue -New Cair...   
405      2  Direct from Owner\n3 bed room flat for sale in...   
406    135  fantastic price for an apartment for sale in Z...   
407    173  3 Rooms + Nanny room with a perfect Price- O-w...   
408     30  Direct from Owner\nسوبر لوكس 170 م بكمباوند ال...   

                                           Summary_url  Summary_Floor_No  \
404  https://aqarmap.com.eg/en/listing/3086496-for-...                 0   
405  https://aqarmap.com.eg/en/listing/3132083-for-...                 4   
406  https://aqarmap.com.eg/en/listing/3112927-for-...                 8   
407  https://aqarmap.com.eg/en/listing/3090510-for-...                 2   
408  https://aqarmap.com.eg/en/listing/2948011-for-...                 5   

     Summary_No_Rooms  Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
404                 3                     3       SUPER_LUX         Garden   
405                 3                     3       SUPER_LUX         230 M²   
406                 2                     2       SUPER_LUX    Main Street   
407                 4                     4       SUPER_LUX         Garden   
408                 3                     2       SUPER_LUX    Main Street   

    Summary_Publishing_Date Summary_Price_EGP  \
404                  170 M²        EG-3086496   
405              EG-3132083        05/01/2022   
406                   97 M²        EG-3112927   
407                  158 M²        EG-3090510   
408                  170 M²        EG-2948011   

                                   Summary_Seller_Type Summary_Price_P_M  
404  Apartment For sale,170M @90th Avenue -New Cair...     2,040,000 EGP  
405  3 bed room flat for sale in Elmerag city in El...    Property Owner  
406  fantastic price for an apartment for sale in Z...     3,644,000 EGP  
407  3 Rooms + Nanny room with a perfect Price- O-west     4,000,000 EGP  
408      سوبر لوكس 170 م بكمباوند البارون سيتى المعادى     1,675,000 EGP

In [27]:
len(df4)

409

In [28]:
fnsh=df4.loc[df4['Summary_Area_m2'].isin(['SEMI_FINISHED', 'EXTRA_SUPER_LUX','SUPER_LUX' ])]
len(fnsh)

fnsh.head()

index                                       Summary_name  \
398    122  Apartment with Garden 180 M² Extra super lux i...   
399     86  أمتداد النرجس, القاهرة الجديدة - التجمع الخامس...   
400     72  Duplex Semi Finished For sale in Bait El Watan...   
401    121  Duplex For Sale in Noor Oasis\nواحة النور, كمب...   
402      1  Apartments For sale in Investors Area\nمنطقة ا...   

                                           Summary_url  Summary_Floor_No  \
398  https://aqarmap.com.eg/en/listing/3115277-for-...                 2   
399  https://aqarmap.com.eg/en/listing/3024135-for-...                 2   
400  https://aqarmap.com.eg/en/listing/3080415-for-...                 1   
401  https://aqarmap.com.eg/en/listing/3057674-for-...                 1   
402  https://aqarmap.com.eg/en/listing/3102080-for-...                 3   

     Summary_No_Rooms  Summary_No_Bathrooms  Summary_Area_m2 Summary_Add_ID  \
398                 3                     3  EXTRA_SUPER_LUX         Garden   
399                 3                     3    SEMI_FINISHED    Main Street   
400                 4                     3    SEMI_FINISHED         Garden   
401                 5                     4    SEMI_FINISHED         Corner   
402                 3                     3    SEMI_FINISHED         Garden   

    Summary_Publishing_Date Summary_Price_EGP  \
398                  180 M²        EG-3115277   
399                  168 M²        EG-3024135   
400                  280 M²        EG-3080415   
401                  340 M²        EG-3057674   
402                  173 M²        EG-3102080   

                                   Summary_Seller_Type Summary_Price_P_M  
398  Apartment with Garden 180 M² Extra super lux i...     4,365,000 EGP  
399                                         27/10/2021     1,529,000 EGP  
400                                         02/12/2021     1,450,000 EGP  
401                      Duplex For Sale in Noor Oasis     1,650,000 EGP  
402                                         16/12/2021     2,000,000 EGP

For the rows that include finishing type values, we can drop them for simplicity since they are 11 only

In [29]:
df5=df4.drop(axis=0, labels=fnsh.index)
df5.tail(5)

index                                       Summary_name  \
393     18  شقة للبيع في العاشر من رمضان الحي ال16\n10th o...   
394     16  شقه للبيع\nAin Shams, Greater Cairo\n90 M² 3 1...   
395    123  شقة للبيع بالحي الخامس عشر دور تالت قسط 430 مت...   
396    107  شقة بالعجمى خطوات من شارع العجمى الرئيسى بالقر...   
397      8  شقه تصلح للتمويل العقاري بقلب حلوان\nGreater C...   

                                           Summary_url  Summary_Floor_No  \
393  https://aqarmap.com.eg/en/listing/3133711-for-...                 1   
394  https://aqarmap.com.eg/en/listing/3133714-for-...                 1   
395  https://aqarmap.com.eg/en/listing/3133362-for-...                 3   
396  https://aqarmap.com.eg/en/listing/3133404-for-...                 2   
397  https://aqarmap.com.eg/en/listing/3133739-for-...                 5   

     Summary_No_Rooms  Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
393                 3                     1           90 M²     EG-3133711   
394                 3                     1           90 M²     EG-3133714   
395                 3                     1           90 M²     EG-3133362   
396                 2                     1           95 M²     EG-3133404   
397                 2                     1           96 M²     EG-3133739   

    Summary_Publishing_Date Summary_Price_EGP  \
393              06/01/2022       235,000 EGP   
394              06/01/2022       300,000 EGP   
395              05/01/2022       220,000 EGP   
396              05/01/2022       240,000 EGP   
397              06/01/2022       390,000 EGP   

                                   Summary_Seller_Type Summary_Price_P_M  
393                                     Property Owner      2,611 EGP/M²  
394                                     Property Owner      3,333 EGP/M²  
395                                     Property Owner      2,444 EGP/M²  
396  شقة بالعجمى خطوات من شارع العجمى الرئيسى بالقر...      2,526 EGP/M²  
397                                     Property Owner      4,063 EGP/M²

In [30]:
len(df5)

398

For the above step, I dropped the rows but found that the length of dataset after dropping them is 385 while we started with 409 and dropped 11 rows only. So I modified this step and made reset index before dropping the rows

Now, I want to see the rows in which the value is a year, to understand where the correct area value is.

In [31]:
years = df5[~df5["Summary_Area_m2"].str.contains('M')]
years.head(2)

index                                       Summary_name  \
61     27  Apartments For sale in Dokki - D169\nالدقي, ال...   
62      3  Apartments For sale in Feini Square - D170\nمي...   

                                          Summary_url  Summary_Floor_No  \
61  https://aqarmap.com.eg/en/listing/3133062-for-...                 5   
62  https://aqarmap.com.eg/en/listing/3133143-for-...                 3   

    Summary_No_Rooms  Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
61                 4                     2            1970            LUX   
62                 2                     2            1980      SUPER_LUX   

   Summary_Publishing_Date Summary_Price_EGP Summary_Seller_Type  \
61             Main Street            145 M²          EG-3133062   
62             Main Street            238 M²          EG-3133143   

   Summary_Price_P_M  
61          5/1/2022  
62          5/1/2022

For the above rows, the values of area is shifted 3 columns, so we need to drop the extra columns.
First, I'll drop these rows, then drop the unneeded columns, then merge the rows back to the dataset

In [32]:
df6=df5.drop(axis=0, index=years.index)
df6.head(2)

index                                       Summary_name  \
0    143  فرصه لا تعوض\nEl Marg, Greater Cairo\n100 M² 3...   
1      5  شقة سوبر لوكس للبييع بمسطرد فرعي من شارع بور س...   

                                         Summary_url  Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133310-for-...                 4   
1  https://aqarmap.com.eg/en/listing/3133765-for-...                 5   

   Summary_No_Rooms  Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
0                 3                     1          100 M²     EG-3133310   
1                 2                     1          100 M²     EG-3133765   

  Summary_Publishing_Date Summary_Price_EGP Summary_Seller_Type  \
0              05/01/2022       300,000 EGP      Property Owner   
1              06/01/2022       300,000 EGP      Property Owner   

  Summary_Price_P_M  
0      3,000 EGP/M²  
1      3,000 EGP/M²

Now Let's drop the unneeded columns

In [33]:
years.drop(axis=1, labels=['Summary_Area_m2','Summary_Add_ID','Summary_Publishing_Date'], inplace=True)
years.head(2)

D:\Anaconda\lib\site-packages\pandas\core\frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


index                                       Summary_name  \
61     27  Apartments For sale in Dokki - D169\nالدقي, ال...   
62      3  Apartments For sale in Feini Square - D170\nمي...   

                                          Summary_url  Summary_Floor_No  \
61  https://aqarmap.com.eg/en/listing/3133062-for-...                 5   
62  https://aqarmap.com.eg/en/listing/3133143-for-...                 3   

    Summary_No_Rooms  Summary_No_Bathrooms Summary_Price_EGP  \
61                 4                     2            145 M²   
62                 2                     2            238 M²   

   Summary_Seller_Type Summary_Price_P_M  
61          EG-3133062          5/1/2022  
62          EG-3133143          5/1/2022

Now, Let's rename the columns with the correct names

In [34]:
years.columns=df5.columns[:-3]
years.head(2)

index                                       Summary_name  \
61     27  Apartments For sale in Dokki - D169\nالدقي, ال...   
62      3  Apartments For sale in Feini Square - D170\nمي...   

                                          Summary_url  Summary_Floor_No  \
61  https://aqarmap.com.eg/en/listing/3133062-for-...                 5   
62  https://aqarmap.com.eg/en/listing/3133143-for-...                 3   

    Summary_No_Rooms  Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
61                 4                     2          145 M²     EG-3133062   
62                 2                     2          238 M²     EG-3133143   

   Summary_Publishing_Date  
61                5/1/2022  
62                5/1/2022

In [35]:
df7=pd.concat([df6,years])
len(df7)

398

Let's have a look on the values of the column now

In [36]:
df7['Summary_Area_m2'].unique()

array(['100 M²', '101 M²', '105 M²', '110 M²', '115 M²', '120 M²',
       '125 M²', '130 M²', '136 M²', '139 M²', '140 M²', '145 M²',
       '150 M²', '156 M²', '160 M²', '164 M²', '165 M²', '170 M²',
       '175 M²', '180 M²', '183 M²', '185 M²', '190 M²', '195 M²',
       '200 M²', '202 M²', '220 M²', '240 M²', '245 M²', '250 M²',
       '265 M²', '300 M²', '330 M²', '500 M²', '530 M²', '56 M²', '57 M²',
       '63 M²', '65 M²', '70 M²', '75 M²', '80 M²', '84 M²', '90 M²',
       '95 M²', '96 M²', '238 M²', '226 M²', '600 M²', '310 M²', '290 M²',
       '108 M²', '168 M²', '198 M²', '135 M²', '155 M²', '138 M²',
       '279 M²', '178 M²', 'EG-2943994', '312 M²', '450 M²', '270 M²',
       'EG-3106020', '230 M²', '113 M²', '295 M²', '323 M²', '277 M²',
       '174 M²', '400 M²', '184 M²', '211 M²', '143 M²', '224 M²',
       '280 M²', '225 M²', '205 M²', '163 M²', '288 M²', '410 M²',
       '166 M²', 'EG-3120640', '154 M²', '182 M²', '149 M²', 'EG-3000957',
       '107 M²', '210 M²', 

It seems that this step returned the value of "Summary_Add_ID" for some rows. This means we should shift them back to the right for 1 step

In [37]:
df8= df7.sort_values(by="Summary_Area_m2").reset_index()
df8.tail(2)

level_0  index                                       Summary_name  \
396      145     74  Apartments For sale in Al Andalus 1\nالأندلس 1...   
397      226     61  Apartments in Compounds in 6 October 185 M² Se...   

                                           Summary_url  Summary_Floor_No  \
396  https://aqarmap.com.eg/en/listing/3120640-for-...                 3   
397  https://aqarmap.com.eg/en/listing/3125808-for-...                 1   

     Summary_No_Rooms  Summary_No_Bathrooms Summary_Area_m2  \
396                 3                     2      EG-3120640   
397                 3                     3      EG-3125808   

                                        Summary_Add_ID  \
396                Apartments For sale in Al Andalus 1   
397  Apartments in Compounds in 6 October 185 M² Se...   

    Summary_Publishing_Date Summary_Price_EGP Summary_Seller_Type  \
396           1,380,000 EGP               NaN                 NaN   
397           1,221,000 EGP               NaN                 NaN   

    Summary_Price_P_M  
396               NaN  
397               NaN

To get these rows, I will filter the column with the values starting with "EG"

In [38]:
ad=df8[df8['Summary_Area_m2'].map(lambda x: x.startswith('EG'))]
ad.head(2)

level_0  index                                       Summary_name  \
381      193    156  Apartments 259 M² Semi Finished in El Bostan\n...   
382      331    100  Apartments 160 M² Semi Finished in El Narges O...   

                                           Summary_url  Summary_Floor_No  \
381  https://aqarmap.com.eg/en/listing/2922190-for-...                 3   
382  https://aqarmap.com.eg/en/listing/2929577-for-...                 1   

     Summary_No_Rooms  Summary_No_Bathrooms Summary_Area_m2  \
381                 3                     3      EG-2922190   
382                 3                     3      EG-2929577   

                                        Summary_Add_ID  \
381       Apartments 259 M² Semi Finished in El Bostan   
382  Apartments 160 M² Semi Finished in El Narges O...   

    Summary_Publishing_Date Summary_Price_EGP Summary_Seller_Type  \
381           2,500,000 EGP               NaN                 NaN   
382           1,515,000 EGP               NaN                 NaN   

    Summary_Price_P_M  
381               NaN  
382               NaN

In [39]:
len(df8)

398

In [40]:
len(ad)

17

Then I will drop these rows, shift the column 1 step to the right, and merge it back with our data

In [41]:
df9=df8.drop(axis=0, labels=ad.index)
df9.head(2)

level_0  index                                       Summary_name  \
0        0    143  فرصه لا تعوض\nEl Marg, Greater Cairo\n100 M² 3...   
1        1      5  شقة سوبر لوكس للبييع بمسطرد فرعي من شارع بور س...   

                                         Summary_url  Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133310-for-...                 4   
1  https://aqarmap.com.eg/en/listing/3133765-for-...                 5   

   Summary_No_Rooms  Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
0                 3                     1          100 M²     EG-3133310   
1                 2                     1          100 M²     EG-3133765   

  Summary_Publishing_Date Summary_Price_EGP Summary_Seller_Type  \
0              05/01/2022       300,000 EGP      Property Owner   
1              06/01/2022       300,000 EGP      Property Owner   

  Summary_Price_P_M  
0      3,000 EGP/M²  
1      3,000 EGP/M²

In [42]:
len(df9)

381

In [43]:
# Now let's insert a new column to shift the data to the right

ad.insert(loc=list(df9.columns).index('Summary_Area_m2'), value=np.nan, column='abc')


In [44]:
ad.head(2)

level_0  index                                       Summary_name  \
381      193    156  Apartments 259 M² Semi Finished in El Bostan\n...   
382      331    100  Apartments 160 M² Semi Finished in El Narges O...   

                                           Summary_url  Summary_Floor_No  \
381  https://aqarmap.com.eg/en/listing/2922190-for-...                 3   
382  https://aqarmap.com.eg/en/listing/2929577-for-...                 1   

     Summary_No_Rooms  Summary_No_Bathrooms  abc Summary_Area_m2  \
381                 3                     3  NaN      EG-2922190   
382                 3                     3  NaN      EG-2929577   

                                        Summary_Add_ID  \
381       Apartments 259 M² Semi Finished in El Bostan   
382  Apartments 160 M² Semi Finished in El Narges O...   

    Summary_Publishing_Date Summary_Price_EGP Summary_Seller_Type  \
381           2,500,000 EGP               NaN                 NaN   
382           1,515,000 EGP               NaN                 NaN   

    Summary_Price_P_M  
381               NaN  
382               NaN

Now Let's rename the columns

In [45]:
new=list(df9.columns)
new.append('dummy')
new

['level_0',
 'index',
 'Summary_name',
 'Summary_url',
 'Summary_Floor_No',
 'Summary_No_Rooms',
 'Summary_No_Bathrooms',
 'Summary_Area_m2',
 'Summary_Add_ID',
 'Summary_Publishing_Date',
 'Summary_Price_EGP',
 'Summary_Seller_Type',
 'Summary_Price_P_M',
 'dummy']

In [46]:
ad.columns= new
ad.columns

Index(['level_0', 'index', 'Summary_name', 'Summary_url', 'Summary_Floor_No',
       'Summary_No_Rooms', 'Summary_No_Bathrooms', 'Summary_Area_m2',
       'Summary_Add_ID', 'Summary_Publishing_Date', 'Summary_Price_EGP',
       'Summary_Seller_Type', 'Summary_Price_P_M', 'dummy'],
      dtype='object')

Now Let's merge both rows

In [47]:
df10=pd.concat([df9, ad.iloc[:,:-2]])
df10.head(2)

level_0  index                                       Summary_name  \
0        0    143  فرصه لا تعوض\nEl Marg, Greater Cairo\n100 M² 3...   
1        1      5  شقة سوبر لوكس للبييع بمسطرد فرعي من شارع بور س...   

                                         Summary_url  Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133310-for-...                 4   
1  https://aqarmap.com.eg/en/listing/3133765-for-...                 5   

   Summary_No_Rooms  Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
0                 3                     1          100 M²     EG-3133310   
1                 2                     1          100 M²     EG-3133765   

  Summary_Publishing_Date Summary_Price_EGP Summary_Seller_Type  \
0              05/01/2022       300,000 EGP      Property Owner   
1              06/01/2022       300,000 EGP      Property Owner   

  Summary_Price_P_M  
0      3,000 EGP/M²  
1      3,000 EGP/M²

In [48]:
len(df10)

398

Now let's check the column values

In [49]:
df10['Summary_Area_m2'].unique()

array(['100 M²', '101 M²', '105 M²', '107 M²', '108 M²', '110 M²',
       '112 M²', '113 M²', '115 M²', '116 M²', '117 M²', '120 M²',
       '121 M²', '125 M²', '126 M²', '127 M²', '130 M²', '131 M²',
       '132 M²', '133 M²', '135 M²', '136 M²', '138 M²', '139 M²',
       '140 M²', '141 M²', '142 M²', '143 M²', '144 M²', '145 M²',
       '147 M²', '149 M²', '150 M²', '152 M²', '153 M²', '154 M²',
       '155 M²', '156 M²', '157 M²', '158 M²', '160 M²', '161 M²',
       '163 M²', '164 M²', '165 M²', '166 M²', '167 M²', '168 M²',
       '169 M²', '170 M²', '171 M²', '172 M²', '173 M²', '174 M²',
       '175 M²', '176 M²', '177 M²', '178 M²', '180 M²', '181 M²',
       '182 M²', '183 M²', '184 M²', '185 M²', '190 M²', '195 M²',
       '198 M²', '200 M²', '202 M²', '203 M²', '205 M²', '206 M²',
       '207 M²', '209 M²', '210 M²', '211 M²', '212 M²', '213 M²',
       '220 M²', '222 M²', '224 M²', '225 M²', '226 M²', '230 M²',
       '236 M²', '238 M²', '240 M²', '245 M²', '248 M²', '250 

We can now move on to the next column "Summary_Add_ID"

In [50]:
df10['Summary_Add_ID'].unique()

array(['EG-3133310', 'EG-3133765', 'EG-3133298',
       'Apartments For Sale in Second Neighborhood',
       'Apartments Semi Finished For Sale in Second Neighborhood',
       'Apartments For sale in Second Neighborhood', 'EG-3133770',
       'EG-3133720', 'EG-1067142',
       'Apartments For sale in Capital Heights', 'EG-3133243',
       'EG-3133432', 'EG-3133512', 'EG-3133745',
       'Apartment for sale, ground floor, in a garden along Al-Thawra Street, in installments over 10 years without interest',
       'EG-3133316', 'EG-3133254', 'شقة لقطة للبيع... المريوطيه فيصل',
       'EG-3133772',
       'شقة للبيع فى بيل في ـ Belle Vie الشيخ زايد الجديدة كمبوند شركة إعمار',
       'Apartment with Garden For sale in Other Neighborhoods In Madenty',
       'EG-3133735',
       'Apartment with garden 116 m for sale in New capital - 5% DP -installments up to 10 years',
       'EG-3125154', 'EG-3133272', 'EG-3133195', 'EG-2891920',
       'EG-3133158', 'EG-3133205', 'EG-3133361',
       'Apar

It seems that some rows include description of the house. Let's see how these rows are organized

In [51]:
ads = df10[~df10["Summary_Add_ID"].str.contains('EG-')]
ads.head(2)

level_0  index                                       Summary_name  \
3      335     36  Apartments For Sale in Second Neighborhood\nSe...   
4      342     21  Apartments Semi Finished For Sale in Second Ne...   

                                         Summary_url  Summary_Floor_No  \
3  https://aqarmap.com.eg/en/listing/2972910-for-...                 4   
4  https://aqarmap.com.eg/en/listing/3010409-for-...                 6   

   Summary_No_Rooms  Summary_No_Bathrooms Summary_Area_m2  \
3                 2                     1          100 M²   
4                 2                     1          100 M²   

                                      Summary_Add_ID Summary_Publishing_Date  \
3         Apartments For Sale in Second Neighborhood              16/09/2021   
4  Apartments Semi Finished For Sale in Second Ne...              17/10/2021   

  Summary_Price_EGP Summary_Seller_Type Summary_Price_P_M  
3               NaN                 NaN               NaN  
4               NaN                 NaN               NaN

In [52]:
len(ads)

181

In [53]:
len(df10)

398

Now I want to know whether these rows have data for the ad ID in a different column or not. So I will check the values in the remaining columns.

In [54]:
ads['Summary_Publishing_Date'].unique()

array(['16/09/2021', '17/10/2021', '02/01/2022', '05/01/2022',
       '19/12/2021', '17/08/2021', '5/1/2022', '22/11/2021', '23/12/2021',
       '02/12/2021', '22/12/2021', '31/07/2021', '12/12/2021',
       '18/12/2021', '28/11/2021', '29/12/2021', '10/10/2021',
       '30/10/2021', '11/12/2021', '13/11/2021', '14/08/2021',
       '28/12/2021', '01/01/2022', '12/07/2021', '30/12/2021',
       '08/12/2021', '23/07/2021', '03/11/2021', '21/12/2021',
       '27/11/2021', '16/11/2021', '17/11/2021', '28/10/2021',
       '09/12/2021', '25/10/2021', '24/11/2021', '16/12/2021',
       '28/08/2021', '06/01/2022', '03/01/2022', '15/12/2021',
       '13/12/2021', '14/12/2021', '04/01/2022', '09/11/2021',
       '25/12/2021', '06/11/2021', '15/08/2021', '01/12/2021',
       '28/07/2021', '04/12/2021', '27/12/2021', '18/11/2021'],
      dtype=object)

The date column values are all dates

In [55]:
ads['Summary_Price_EGP'].unique()

array([nan], dtype=object)

In [56]:
ads['Summary_Price_P_M'].unique()

array([nan], dtype=object)

In [57]:
ads['Summary_Seller_Type'].unique()

array([nan], dtype=object)

So these rows don't have ad ID. we can replace the values with "NA"

In [58]:
ads["Summary_Add_ID"]="NA"

<ipython-input-58-1f8447dc9180>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ads["Summary_Add_ID"]="NA"


Now Let's replace these rows in the dataset

In [59]:
df11=df10.drop(axis=0, labels=ads.index)

len(df11)

217

In [60]:
df12=pd.concat([df11,ads])
df12.head(2)

level_0  index                                       Summary_name  \
0        0    143  فرصه لا تعوض\nEl Marg, Greater Cairo\n100 M² 3...   
1        1      5  شقة سوبر لوكس للبييع بمسطرد فرعي من شارع بور س...   

                                         Summary_url  Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133310-for-...                 4   
1  https://aqarmap.com.eg/en/listing/3133765-for-...                 5   

   Summary_No_Rooms  Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
0                 3                     1          100 M²     EG-3133310   
1                 2                     1          100 M²     EG-3133765   

  Summary_Publishing_Date Summary_Price_EGP Summary_Seller_Type  \
0              05/01/2022       300,000 EGP      Property Owner   
1              06/01/2022       300,000 EGP      Property Owner   

  Summary_Price_P_M  
0      3,000 EGP/M²  
1      3,000 EGP/M²

In [61]:
len(df12)

398

Now let's check the next column

In [62]:
df12['Summary_Publishing_Date'].unique()

array(['05/01/2022', '06/01/2022', '1,433,800 EGP', '31/12/2021',
       '10/07/2021', '5/1/2022', '1,945,800 EGP', '20/12/2021',
       '02/01/2022', '27/10/2021', '12/12/2021', '06/07/2021',
       '21/12/2021', '25/12/2021', '22/12/2021', '13/12/2021',
       '27/12/2021', '29/12/2021', '04/01/2022', '09/11/2021',
       '10/11/2021', '28/12/2021', '06/08/2021', '01/08/2021',
       '03/01/2022', '26/07/2021', '14/12/2021', '04/11/2021',
       '11/11/2021', '22/11/2021', '11/07/2021', '12/09/2021',
       '29/07/2021', '1,490,110 EGP',
       'Apartments 259 M² Semi Finished in El Bostan',
       'Apartments 160 M² Semi Finished in El Narges Omarat',
       '24/08/2021',
       'Duplex 166 M² For Sale in Atika Compound - New Plan',
       '09/10/2021', '08/11/2021',
       'Apartments 150 M² Semi Finished in New El Andalous',
       'Apartments Semi Finished For Sale in New Heliopolis',
       'Apartment with Garden in New Heliopolis 162 M² Semi Finished For Sale',
       '19/12/20

Let's see the rows that include values other than the date

In [63]:
dates=df12[~df12["Summary_Publishing_Date"].str.contains('/20')]
len(dates)

12

In [64]:
dates

level_0  index                                       Summary_name  \
8        160      9  Apartment 107 m for sale in Kattameya Gate apa...   
58       167      8  Apartment 138 m for sale in Kattameya Gate, th...   
368      341      3  Duplex for sale 83m with garden in Monte Napol...   
381      193    156  Apartments 259 M² Semi Finished in El Bostan\n...   
382      331    100  Apartments 160 M² Semi Finished in El Narges O...   
384      195     24  Duplex 166 M² For Sale in Atika Compound - New...   
388      190     98  Apartments 150 M² Semi Finished in New El Anda...   
389      264    127  Apartments Semi Finished For Sale in New Helio...   
390      272    144  Apartment with Garden in New Heliopolis 162 M²...   
395      219     70  Apartments For sale in Al Andalus 1\nالأندلس 1...   
396      145     74  Apartments For sale in Al Andalus 1\nالأندلس 1...   
397      226     61  Apartments in Compounds in 6 October 185 M² Se...   

                                           Summary_url  Summary_Floor_No  \
8    https://aqarmap.com.eg/en/listing/1067142-proj...                 2   
58   https://aqarmap.com.eg/en/listing/1067133-proj...                 2   
368  https://aqarmap.com.eg/en/listing/2958379-proj...                 0   
381  https://aqarmap.com.eg/en/listing/2922190-for-...                 3   
382  https://aqarmap.com.eg/en/listing/2929577-for-...                 1   
384  https://aqarmap.com.eg/en/listing/2953270-for-...                 2   
388  https://aqarmap.com.eg/en/listing/3089009-for-...                 3   
389  https://aqarmap.com.eg/en/listing/3099488-for-...                 5   
390  https://aqarmap.com.eg/en/listing/3099507-for-...                 0   
395  https://aqarmap.com.eg/en/listing/3120346-for-...                 2   
396  https://aqarmap.com.eg/en/listing/3120640-for-...                 3   
397  https://aqarmap.com.eg/en/listing/3125808-for-...                 1   

     Summary_No_Rooms  Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
8                   2                     2          107 M²     EG-1067142   
58                  4                     2          138 M²     EG-1067133   
368                 1                     1           83 M²     EG-2958379   
381                 3                     3             NaN     EG-2922190   
382                 3                     3             NaN     EG-2929577   
384                 3                     2             NaN     EG-2953270   
388                 3                     2             NaN     EG-3089009   
389                 3                     2             NaN     EG-3099488   
390                 3                     3             NaN     EG-3099507   
395                 3                     2             NaN     EG-3120346   
396                 3                     2             NaN     EG-3120640   
397                 3                     3             NaN     EG-3125808   

                               Summary_Publishing_Date Summary_Price_EGP  \
8                                        1,433,800 EGP               NaN   
58                                       1,945,800 EGP               NaN   
368                                      1,490,110 EGP               NaN   
381       Apartments 259 M² Semi Finished in El Bostan     2,500,000 EGP   
382  Apartments 160 M² Semi Finished in El Narges O...     1,515,000 EGP   
384  Duplex 166 M² For Sale in Atika Compound - New...     2,324,000 EGP   
388  Apartments 150 M² Semi Finished in New El Anda...     1,300,000 EGP   
389  Apartments Semi Finished For Sale in New Helio...       968,000 EGP   
390  Apartment with Garden in New Heliopolis 162 M²...       862,000 EGP   
395                Apartments For sale in Al Andalus 1     1,317,500 EGP   
396                Apartments For sale in Al Andalus 1     1,380,000 EGP   
397  Apartments in Compounds in 6 October 185 M² Se...     1,221,000 EGP   

    Summary_Seller_Type Summary_Price_P_M  
8    

The first 3 rows should be shifted 1 step to the right to display the price value correctly. And it doesn't include the dates in any other column so we can replace the values with NA. For the remaining ones, we can only replace the values with NA without shifting because the next column  (Price) is correctly located.

In [65]:
dates.iloc[0:3]['Summary_Price_EGP']=dates.iloc[0:3]['Summary_Publishing_Date']
dates['Summary_Publishing_Date']="NA"
dates.head(4)

<ipython-input-65-ca4c1b66385a>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dates.iloc[0:3]['Summary_Price_EGP']=dates.iloc[0:3]['Summary_Publishing_Date']
<ipython-input-65-ca4c1b66385a>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dates['Summary_Publishing_Date']="NA"


level_0  index                                       Summary_name  \
8        160      9  Apartment 107 m for sale in Kattameya Gate apa...   
58       167      8  Apartment 138 m for sale in Kattameya Gate, th...   
368      341      3  Duplex for sale 83m with garden in Monte Napol...   
381      193    156  Apartments 259 M² Semi Finished in El Bostan\n...   

                                           Summary_url  Summary_Floor_No  \
8    https://aqarmap.com.eg/en/listing/1067142-proj...                 2   
58   https://aqarmap.com.eg/en/listing/1067133-proj...                 2   
368  https://aqarmap.com.eg/en/listing/2958379-proj...                 0   
381  https://aqarmap.com.eg/en/listing/2922190-for-...                 3   

     Summary_No_Rooms  Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
8                   2                     2          107 M²     EG-1067142   
58                  4                     2          138 M²     EG-1067133   
368                 1                     1           83 M²     EG-2958379   
381                 3                     3             NaN     EG-2922190   

    Summary_Publishing_Date Summary_Price_EGP Summary_Seller_Type  \
8                        NA     1,433,800 EGP                 NaN   
58                       NA     1,945,800 EGP                 NaN   
368                      NA     1,490,110 EGP                 NaN   
381                      NA     2,500,000 EGP                 NaN   

    Summary_Price_P_M  
8                 NaN  
58                NaN  
368               NaN  
381               NaN

Now Let's drop the old dates rows and then merge the updated ones

In [66]:
df13=df12.drop(axis=0, labels=dates.index)
df14=pd.concat([df13,dates])
df13.head(2)

level_0  index                                       Summary_name  \
0        0    143  فرصه لا تعوض\nEl Marg, Greater Cairo\n100 M² 3...   
1        1      5  شقة سوبر لوكس للبييع بمسطرد فرعي من شارع بور س...   

                                         Summary_url  Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133310-for-...                 4   
1  https://aqarmap.com.eg/en/listing/3133765-for-...                 5   

   Summary_No_Rooms  Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
0                 3                     1          100 M²     EG-3133310   
1                 2                     1          100 M²     EG-3133765   

  Summary_Publishing_Date Summary_Price_EGP Summary_Seller_Type  \
0              05/01/2022       300,000 EGP      Property Owner   
1              06/01/2022       300,000 EGP      Property Owner   

  Summary_Price_P_M  
0      3,000 EGP/M²  
1      3,000 EGP/M²

In [67]:
len(df14)

398

Now Let's take another look on the dataset

In [68]:
df14.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 398 entries, 0 to 397
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   level_0                  398 non-null    int64 
 1   index                    398 non-null    int64 
 2   Summary_name             398 non-null    object
 3   Summary_url              398 non-null    object
 4   Summary_Floor_No         398 non-null    int32 
 5   Summary_No_Rooms         398 non-null    int32 
 6   Summary_No_Bathrooms     398 non-null    int32 
 7   Summary_Area_m2          381 non-null    object
 8   Summary_Add_ID           398 non-null    object
 9   Summary_Publishing_Date  398 non-null    object
 10  Summary_Price_EGP        112 non-null    object
 11  Summary_Seller_Type      92 non-null     object
 12  Summary_Price_P_M        92 non-null     object
dtypes: int32(3), int64(2), object(8)
memory usage: 38.9+ KB


At this stage, it's disappointing to see that the price is missing from around 75% of the data, which means that if we were planning to predict the price then the accuracy will be low because we can't impute such a large amount of data and rely on it to train our model. But I'll leave the column for now to use it in analysis. I'll also take a look at some of these rows in the first version of our data for health check to make sure the prices were missing from the beginning and ensure I didn't delete them by mistake during preprocessing.

In [69]:
# First, let's retrieve the rows with null values. But we can't use their index because we used reset index in previous steps
miss=df14[(df14['Summary_Price_EGP']).isnull()]['Summary_url']
miss.head()

2     https://aqarmap.com.eg/en/listing/3133298-for-...
10    https://aqarmap.com.eg/en/listing/3133243-for-...
23    https://aqarmap.com.eg/en/listing/3125154-for-...
26    https://aqarmap.com.eg/en/listing/2891920-for-...
61    https://aqarmap.com.eg/en/listing/3108599-for-...
Name: Summary_url, dtype: object

In [70]:
# Now let's merge the above data with df1 the original dataset
miss2=pd.merge(df1, miss, how='inner')
miss2.head(2)

Summary_name  \
0  شقة 270 متر في كمبوند حدائق الفدا نص تشطيب دور...   
1  شقة 200 متر في فيلات الدفاع الوطني 75 % تشطيب+...   

                                         Summary_url Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133688-for-...                1   
1  https://aqarmap.com.eg/en/listing/3133659-for-...                2   

  Summary_No_Rooms Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
0                3                    3            2015  SEMI_FINISHED   
1                3                    3            2015            LUX   

  Summary_Publishing_Date Summary_Price_EGP  \
0             Side Street            270 M²   
1             Side Street            200 M²   

                                 Summary_Seller_Type Summary_Price_P_M  
0  شقة 270 متر في كمبوند حدائق الفدا نص تشطيب دور...        05/01/2022  
1  شقة 200 متر في فيلات الدفاع الوطني 75 % تشطيب+...        05/01/2022

In [71]:
[miss2[col].str.count('EGP').sum() for col in df1.columns]

[297, 0, 0, 0.0, 0, 0, 0, 0, 0, 11, 0]

In [72]:
df1.columns

Index(['Summary_name', 'Summary_url', 'Summary_Floor_No', 'Summary_No_Rooms',
       'Summary_No_Bathrooms', 'Summary_Area_m2', 'Summary_Add_ID',
       'Summary_Publishing_Date', 'Summary_Price_EGP', 'Summary_Seller_Type',
       'Summary_Price_P_M'],
      dtype='object')

In [73]:
len(miss2)

286

Good news here! The first column "Summary_name" has the required price data. But it might be strange that the values are 297 although the # rows are 286 in miss2. Let's dive deeper.

In [74]:
df1['Summary_name'][0].values

array(['شقة للبيع مكان مميز جدااا\n6th of October, Greater Cairo\n245 M² 4 2\n2,000,000 EGP',
       'Corner Penthouse Apartment in Cairo University Compound Sheikh Zayed 4, 6th of October\nطريق وصلة دهشور, 6 أكتوبر\n350 M² 4 4\n1,250,000 EGP',
       'Apartments For sale in Second Neighborhood\nالشطر الثاني, زهراء المعادي\n100 M² 2 1\n1,050,000 EGP'],
      dtype=object)

In [75]:
df1.head(1)

Summary_name  \
0  شقة للبيع مكان مميز جدااا\n6th of October, Gre...   

                                         Summary_url Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133778-for-...                4   

  Summary_No_Rooms Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
0                4                    2          245 M²     EG-3133778   

  Summary_Publishing_Date Summary_Price_EGP Summary_Seller_Type  \
0              06/01/2022     2,000,000 EGP      Property Owner   

  Summary_Price_P_M  
0      8,163 EGP/M²

It seems that the summary includes multiple prices, so we can't rely on it for extracting the price. Also for the "price/m"  column, it is also missing the needed price/m values in the rows with missing values. So we will keep the price column as is.

Now all columns should include consistent data shapes!

For the column "Summary_Seller_Type", I think the information it holds is not important, and it has alot of missing and incorrect values, so we can drop it. Also "Summary_Price_P_M" is redundent, so I will drop it as well.

In [76]:
df15=df14.drop(axis=1,labels=['Summary_Seller_Type','Summary_Price_P_M'] )
df15.head(2)

level_0  index                                       Summary_name  \
0        0    143  فرصه لا تعوض\nEl Marg, Greater Cairo\n100 M² 3...   
1        1      5  شقة سوبر لوكس للبييع بمسطرد فرعي من شارع بور س...   

                                         Summary_url  Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133310-for-...                 4   
1  https://aqarmap.com.eg/en/listing/3133765-for-...                 5   

   Summary_No_Rooms  Summary_No_Bathrooms Summary_Area_m2 Summary_Add_ID  \
0                 3                     1          100 M²     EG-3133310   
1                 2                     1          100 M²     EG-3133765   

  Summary_Publishing_Date Summary_Price_EGP  
0              05/01/2022       300,000 EGP  
1              06/01/2022       300,000 EGP

For the "Summary_Area_m2" column, we can change the values to the area numeric value without the M2 and change the type to float to facilitate using it in our analysis

In [77]:
unit=df15['Summary_Area_m2'][1].split()[1]
unit

'M²'

In [78]:
df15['Summary_Area_m2']=df15['Summary_Area_m2'].str.replace(unit,"").astype('float')

df15.head(2)

level_0  index                                       Summary_name  \
0        0    143  فرصه لا تعوض\nEl Marg, Greater Cairo\n100 M² 3...   
1        1      5  شقة سوبر لوكس للبييع بمسطرد فرعي من شارع بور س...   

                                         Summary_url  Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133310-for-...                 4   
1  https://aqarmap.com.eg/en/listing/3133765-for-...                 5   

   Summary_No_Rooms  Summary_No_Bathrooms  Summary_Area_m2 Summary_Add_ID  \
0                 3                     1            100.0     EG-3133310   
1                 2                     1            100.0     EG-3133765   

  Summary_Publishing_Date Summary_Price_EGP  
0              05/01/2022       300,000 EGP  
1              06/01/2022       300,000 EGP

We can also cleanup the price column by removing the "EGP" to facilitate using the numbers in statistics

In [79]:
df15['Summary_Price_EGP']=df15['Summary_Price_EGP'].str.replace(" EGP","")
df15.head(2)

level_0  index                                       Summary_name  \
0        0    143  فرصه لا تعوض\nEl Marg, Greater Cairo\n100 M² 3...   
1        1      5  شقة سوبر لوكس للبييع بمسطرد فرعي من شارع بور س...   

                                         Summary_url  Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133310-for-...                 4   
1  https://aqarmap.com.eg/en/listing/3133765-for-...                 5   

   Summary_No_Rooms  Summary_No_Bathrooms  Summary_Area_m2 Summary_Add_ID  \
0                 3                     1            100.0     EG-3133310   
1                 2                     1            100.0     EG-3133765   

  Summary_Publishing_Date Summary_Price_EGP  
0              05/01/2022           300,000  
1              06/01/2022           300,000

Now, we can double check the "Summary URL" column for any useful data.

In [80]:
print(df15['Summary_name'][3])
print("*******")
print(df15['Summary_name'][7])
print("*******")
print(df15['Summary_name'][9])
print("*******")
print(df15['Summary_name'][77])

Apartments For Sale in Second Neighborhood
Second Neighborhood, Zahraa El Maadi
100 M² 2 1
1,050,000 EGP
*******
شقة في حسن محمد - فيصل
Faisal, Greater Cairo
105 M² 2 1
350,000 EGP
*******
Apartments For sale in Capital Heights
كابيتال هايتس, كمبوند بالعاصمة الادارية الجديدة
107 M² 2 2
2,663,460 EGP
*******
Apartment for sale in New Capital Scene7 Compound -5 Years Installments
كمبوند سين 7 - اكام, كمبوند بالعاصمة الادارية الجديدة
144 M² 3 2
2,129,717 EGP


Although the area name is mentioned in the URL column, but it doesn't follow a unified sequence. For example, in two ads, it was written "كمبوند بالعاصمة الادارية الجديدة" not the area name directly. For the first & second entries above, Zahraa el Maadi is the second item in the second row, while Faisal is the first entry in the second row. which makes it hard to extract this area name.

But it seems that the price & area are mentioned in the URL, so we can extract them and fill in the missing values.

Let's start with the area. We will add a new column including the area extracted from URL

In [81]:
df16=df15.copy()

In [82]:
df16['Area']=df16['Summary_name'].apply(lambda x: x.split('\n')[2])
df16.head(2)

level_0  index                                       Summary_name  \
0        0    143  فرصه لا تعوض\nEl Marg, Greater Cairo\n100 M² 3...   
1        1      5  شقة سوبر لوكس للبييع بمسطرد فرعي من شارع بور س...   

                                         Summary_url  Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133310-for-...                 4   
1  https://aqarmap.com.eg/en/listing/3133765-for-...                 5   

   Summary_No_Rooms  Summary_No_Bathrooms  Summary_Area_m2 Summary_Add_ID  \
0                 3                     1            100.0     EG-3133310   
1                 2                     1            100.0     EG-3133765   

  Summary_Publishing_Date Summary_Price_EGP        Area  
0              05/01/2022           300,000  100 M² 3 1  
1              06/01/2022           300,000  100 M² 2 1

Let's specify the rows with missing "Area" values

In [83]:
no_area1=df16[df16['Summary_Area_m2'].isna()]
no_area1.head(2)

level_0  index                                       Summary_name  \
383      101     33  Apartments Semi Finished For Sale in Southern ...   
385      158     29  Direct from Owner\nGREAT Opportunity | Apartme...   

                                           Summary_url  Summary_Floor_No  \
383  https://aqarmap.com.eg/en/listing/2943994-for-...                 2   
385  https://aqarmap.com.eg/en/listing/3000957-for-...                 2   

     Summary_No_Rooms  Summary_No_Bathrooms  Summary_Area_m2 Summary_Add_ID  \
383                 3                     2              NaN     EG-2943994   
385                 2                     2              NaN     EG-3000957   

    Summary_Publishing_Date Summary_Price_EGP  \
383              24/08/2021         2,170,000   
385              09/10/2021         1,700,000   

                                                  Area  
383                                         155 M² 3 2  
385  Palm Springs Residential Community - Alex Dese...

In [84]:
no_area2=df16[df16['Summary_Area_m2']=="NA"]
no_area2

Empty DataFrame
Columns: [level_0, index, Summary_name, Summary_url, Summary_Floor_No, Summary_No_Rooms, Summary_No_Bathrooms, Summary_Area_m2, Summary_Add_ID, Summary_Publishing_Date, Summary_Price_EGP, Area]
Index: []

Then we will work on the rows with nan values

In [85]:
no_area1.sort_values(by='Area',inplace=True)
no_area1['Area']=no_area1['Area'].apply(lambda x: x.split(' ')[0])
no_area1.head(2)

<ipython-input-85-b7b49dcb9dce>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  no_area1.sort_values(by='Area',inplace=True)
<ipython-input-85-b7b49dcb9dce>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  no_area1['Area']=no_area1['Area'].apply(lambda x: x.split(' ')[0])


level_0  index                                       Summary_name  \
389      264    127  Apartments Semi Finished For Sale in New Helio...   
394      269    133  امتلك شقتك في قلب المعادي بجوار نادي وادي دجله...   

                                           Summary_url  Summary_Floor_No  \
389  https://aqarmap.com.eg/en/listing/3099488-for-...                 5   
394  https://aqarmap.com.eg/en/listing/3119088-for-...                 3   

     Summary_No_Rooms  Summary_No_Bathrooms  Summary_Area_m2 Summary_Add_ID  \
389                 3                     2              NaN     EG-3099488   
394                 3                     2              NaN     EG-3119088   

    Summary_Publishing_Date Summary_Price_EGP Area  
389                      NA           968,000  144  
394              27/12/2021         1,187,000  145

In [86]:
no_area1['Area'].unique()

array(['144', '145', '150', '155', '160', '162', '166', '174', '175',
       '184', '185', '258', '259', 'Palm'], dtype=object)

We only have an incorrect value "Palm". Let's drop this row and update the area value in the original column

In [87]:
# Writing the value of palm by hand might not be the best practice, because we should use automated process and avoid hard 
# coding so that we could maintain the code re-producability with other datasets. But I'm using it due to time limitation.

ind=no_area1[no_area1['Area']== "Palm"].index[0]
ind

385

In [88]:
no_area3=no_area1.drop(axis=0, index=ind)
no_area3['Summary_Area_m2']=no_area3['Area']
no_area3['Summary_Area_m2'].unique()

array(['144', '145', '150', '155', '160', '162', '166', '174', '175',
       '184', '185', '258', '259'], dtype=object)

We can now drop the old rows and add the updated ones, also let's change the type to float

In [89]:
df17=df16.drop(axis=0, index=no_area3.index)
df18=pd.concat([df17, no_area3])
df18.drop(axis=0, index=ind, inplace=True)
df18['Summary_Area_m2']=df18['Summary_Area_m2'].astype('float')
len(df18)

397

Now let's work on the same steps to update the price value

In [90]:
df18['Price']=df18['Summary_name'].apply(lambda x: x.split('\n')[-1])
df18.head(2)

level_0  index                                       Summary_name  \
0        0    143  فرصه لا تعوض\nEl Marg, Greater Cairo\n100 M² 3...   
1        1      5  شقة سوبر لوكس للبييع بمسطرد فرعي من شارع بور س...   

                                         Summary_url  Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133310-for-...                 4   
1  https://aqarmap.com.eg/en/listing/3133765-for-...                 5   

   Summary_No_Rooms  Summary_No_Bathrooms  Summary_Area_m2 Summary_Add_ID  \
0                 3                     1            100.0     EG-3133310   
1                 2                     1            100.0     EG-3133765   

  Summary_Publishing_Date Summary_Price_EGP        Area        Price  
0              05/01/2022           300,000  100 M² 3 1  300,000 EGP  
1              06/01/2022           300,000  100 M² 2 1  300,000 EGP

In [91]:
print(len(df18['Price']))
df18['Price'].str.count('EGP').sum()

397


397

Let's check the null and NaN values in the original "'Summary_Price_EGP'" column 

In [92]:
no_prc1=df18[df18['Summary_Price_EGP'].isna()]
len(no_prc1)

286

In [93]:
no_prc2=df18[df18['Summary_Price_EGP']=="NA"]
len(no_prc2)

0

In [94]:
no_prc1['Summary_Price_EGP']=no_prc1['Price']
no_prc1.head(2)

<ipython-input-94-9e25ec334d60>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  no_prc1['Summary_Price_EGP']=no_prc1['Price']


level_0  index                                       Summary_name  \
2       177    122  Apartments For sale in Kafr Nassar\nكفر نصار, ...   
10       77    127  Apartments For sale in Dokki - D174\nالدقي, ال...   

                                          Summary_url  Summary_Floor_No  \
2   https://aqarmap.com.eg/en/listing/3133298-for-...                11   
10  https://aqarmap.com.eg/en/listing/3133243-for-...                 8   

    Summary_No_Rooms  Summary_No_Bathrooms  Summary_Area_m2 Summary_Add_ID  \
2                  3                     1            100.0     EG-3133298   
10                 2                     2            108.0     EG-3133243   

   Summary_Publishing_Date Summary_Price_EGP        Area          Price  
2               05/01/2022       255,000 EGP  100 M² 3 1    255,000 EGP  
10              05/01/2022     1,500,000 EGP  108 M² 2 2  1,500,000 EGP

In [95]:
df18.drop(axis=0, index=no_prc1.index, inplace=True)
df18.head(2)

level_0  index                                       Summary_name  \
0        0    143  فرصه لا تعوض\nEl Marg, Greater Cairo\n100 M² 3...   
1        1      5  شقة سوبر لوكس للبييع بمسطرد فرعي من شارع بور س...   

                                         Summary_url  Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133310-for-...                 4   
1  https://aqarmap.com.eg/en/listing/3133765-for-...                 5   

   Summary_No_Rooms  Summary_No_Bathrooms  Summary_Area_m2 Summary_Add_ID  \
0                 3                     1            100.0     EG-3133310   
1                 2                     1            100.0     EG-3133765   

  Summary_Publishing_Date Summary_Price_EGP        Area        Price  
0              05/01/2022           300,000  100 M² 3 1  300,000 EGP  
1              06/01/2022           300,000  100 M² 2 1  300,000 EGP

In [96]:
df19=pd.concat([df18,no_prc1])
len(df19)

397

Let's clean the price column by removing the "EGP" and changing the type to integer

In [97]:
df19['Summary_Price_EGP']=df19['Summary_Price_EGP'].str.replace(" EGP","")
df19['Summary_Price_EGP']=df19['Summary_Price_EGP'].apply(lambda x: x.replace(',',""))
df19['Summary_Price_EGP']=df19['Summary_Price_EGP'].apply(lambda x: int(x))

df19.head(2)

level_0  index                                       Summary_name  \
0        0    143  فرصه لا تعوض\nEl Marg, Greater Cairo\n100 M² 3...   
1        1      5  شقة سوبر لوكس للبييع بمسطرد فرعي من شارع بور س...   

                                         Summary_url  Summary_Floor_No  \
0  https://aqarmap.com.eg/en/listing/3133310-for-...                 4   
1  https://aqarmap.com.eg/en/listing/3133765-for-...                 5   

   Summary_No_Rooms  Summary_No_Bathrooms  Summary_Area_m2 Summary_Add_ID  \
0                 3                     1            100.0     EG-3133310   
1                 2                     1            100.0     EG-3133765   

  Summary_Publishing_Date  Summary_Price_EGP        Area        Price  
0              05/01/2022             300000  100 M² 3 1  300,000 EGP  
1              06/01/2022             300000  100 M² 2 1  300,000 EGP

In [98]:
df19.isnull().sum()

level_0                    0
index                      0
Summary_name               0
Summary_url                0
Summary_Floor_No           0
Summary_No_Rooms           0
Summary_No_Bathrooms       0
Summary_Area_m2            0
Summary_Add_ID             0
Summary_Publishing_Date    0
Summary_Price_EGP          0
Area                       0
Price                      0
dtype: int64

In [99]:
[len(df19[df19[col]=='NA']) for col in df19.columns]

[0, 0, 0, 0, 0, 0, 0, 0, 181, 12, 0, 0, 0]

In [100]:
df19.columns

Index(['level_0', 'index', 'Summary_name', 'Summary_url', 'Summary_Floor_No',
       'Summary_No_Rooms', 'Summary_No_Bathrooms', 'Summary_Area_m2',
       'Summary_Add_ID', 'Summary_Publishing_Date', 'Summary_Price_EGP',
       'Area', 'Price'],
      dtype='object')

It's worth mentioning that the "Summary_Add_ID" column doesn't include useful data that is related to the house, so we can drop it. But before we do so, we can use it to impute the missing values in the date column, because the more recent the ad was, the larger its number would be

In [101]:
df20=df19.sort_values(by='Summary_Add_ID')
df20.head(2)

level_0  index                                       Summary_name  \
58      167      8  Apartment 138 m for sale in Kattameya Gate, th...   
8       160      9  Apartment 107 m for sale in Kattameya Gate apa...   

                                          Summary_url  Summary_Floor_No  \
58  https://aqarmap.com.eg/en/listing/1067133-proj...                 2   
8   https://aqarmap.com.eg/en/listing/1067142-proj...                 2   

    Summary_No_Rooms  Summary_No_Bathrooms  Summary_Area_m2 Summary_Add_ID  \
58                 4                     2            138.0     EG-1067133   
8                  2                     2            107.0     EG-1067142   

   Summary_Publishing_Date  Summary_Price_EGP        Area          Price  
58                      NA            1945800  138 M² 4 2  1,945,800 EGP  
8                       NA            1433800  107 M² 2 2  1,433,800 EGP

In [102]:
df20['Summary_Publishing_Date'].replace("NA", np.nan, inplace=True)

In [103]:
df20['Summary_Publishing_Date'].fillna( method="ffill", inplace=True)
df20.head(3)

level_0  index                                       Summary_name  \
58       167      8  Apartment 138 m for sale in Kattameya Gate, th...   
8        160      9  Apartment 107 m for sale in Kattameya Gate apa...   
109      255    132  Apartments Semi Finished For Sale in New Helio...   

                                           Summary_url  Summary_Floor_No  \
58   https://aqarmap.com.eg/en/listing/1067133-proj...                 2   
8    https://aqarmap.com.eg/en/listing/1067142-proj...                 2   
109  https://aqarmap.com.eg/en/listing/2888841-for-...                 0   

     Summary_No_Rooms  Summary_No_Bathrooms  Summary_Area_m2 Summary_Add_ID  \
58                  4                     2            138.0     EG-1067133   
8                   2                     2            107.0     EG-1067142   
109                 3                     3            150.0     EG-2888841   

    Summary_Publishing_Date  Summary_Price_EGP        Area          Price  
58                      NaN            1945800  138 M² 4 2  1,945,800 EGP  
8                       NaN            1433800  107 M² 2 2  1,433,800 EGP  
109              06/07/2021             774000  150 M² 3 3    774,000 EGP

It seems that imputation didn't work on the first 2 rows because it didn't find their previous values to use. So we can impute again using bfill argument.

In [104]:
df20['Summary_Publishing_Date'].fillna( method="bfill", inplace=True)
# Let's also change the value to date time format to facilitate the analysis
df20['Summary_Publishing_Date']=df20['Summary_Publishing_Date'].apply(func=pd.to_datetime)
df20.head(3)

level_0  index                                       Summary_name  \
58       167      8  Apartment 138 m for sale in Kattameya Gate, th...   
8        160      9  Apartment 107 m for sale in Kattameya Gate apa...   
109      255    132  Apartments Semi Finished For Sale in New Helio...   

                                           Summary_url  Summary_Floor_No  \
58   https://aqarmap.com.eg/en/listing/1067133-proj...                 2   
8    https://aqarmap.com.eg/en/listing/1067142-proj...                 2   
109  https://aqarmap.com.eg/en/listing/2888841-for-...                 0   

     Summary_No_Rooms  Summary_No_Bathrooms  Summary_Area_m2 Summary_Add_ID  \
58                  4                     2            138.0     EG-1067133   
8                   2                     2            107.0     EG-1067142   
109                 3                     3            150.0     EG-2888841   

    Summary_Publishing_Date  Summary_Price_EGP        Area          Price  
58               2021-06-07            1945800  138 M² 4 2  1,945,800 EGP  
8                2021-06-07            1433800  107 M² 2 2  1,433,800 EGP  
109              2021-06-07             774000  150 M² 3 3    774,000 EGP

The final sanity check..

In [105]:
df20.isnull().sum()

level_0                    0
index                      0
Summary_name               0
Summary_url                0
Summary_Floor_No           0
Summary_No_Rooms           0
Summary_No_Bathrooms       0
Summary_Area_m2            0
Summary_Add_ID             0
Summary_Publishing_Date    0
Summary_Price_EGP          0
Area                       0
Price                      0
dtype: int64

In [106]:
[len(df20[df20[col]=='NA']) for col in df20.columns]

[0, 0, 0, 0, 0, 0, 0, 0, 181, 0, 0, 0, 0]

Good. Now all columns contain values. Let's drop the unneeded columns. 

In [107]:
df21=df20.iloc[:, 4:-2]
df21.drop(axis=1, labels='Summary_Add_ID', inplace=True)
df21.head(2)

Summary_Floor_No  Summary_No_Rooms  Summary_No_Bathrooms  Summary_Area_m2  \
58                 2                 4                     2            138.0   
8                  2                 2                     2            107.0   

   Summary_Publishing_Date  Summary_Price_EGP  
58              2021-06-07            1945800  
8               2021-06-07            1433800

In [108]:
df21.reset_index(inplace=True)

In [109]:
df21.drop(inplace=True, axis=1, labels='index')

In [110]:
df21.head()

Summary_Floor_No  Summary_No_Rooms  Summary_No_Bathrooms  Summary_Area_m2  \
0                 2                 4                     2            138.0   
1                 2                 2                     2            107.0   
2                 0                 3                     3            150.0   
3                 0                 3                     3            175.0   
4                 6                 3                     2            120.0   

  Summary_Publishing_Date  Summary_Price_EGP  
0              2021-06-07            1945800  
1              2021-06-07            1433800  
2              2021-06-07             774000  
3              2021-06-07             896000  
4              2021-10-07             396000

In [111]:
df21.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 397 entries, 0 to 396
Data columns (total 6 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   Summary_Floor_No         397 non-null    int32         
 1   Summary_No_Rooms         397 non-null    int32         
 2   Summary_No_Bathrooms     397 non-null    int32         
 3   Summary_Area_m2          397 non-null    float64       
 4   Summary_Publishing_Date  397 non-null    datetime64[ns]
 5   Summary_Price_EGP        397 non-null    int64         
dtypes: datetime64[ns](1), float64(1), int32(3), int64(1)
memory usage: 14.1 KB


Lesson learned: I could have saved more time by investigating the "Summary" column from the beginning and extracting the missing values from it.

Now, our data is ready for the analysis!

#### Analyzing the data
(Ahmed)